### PR TITLE
Generate Algebird microsite via sbt-microsites

### PR DESCRIPTION
**Pull request by sritchie**
This PR adds a new, beautiful (if I do say so myself!) microsite for Algebird using the [SBT-Microsites plugin](https://github.com/47deg/sbt-microsites).

I've also converted the Algebird wiki into an executable set of markdown files using the wonderful [tut](https://github.com/tpolecat/tut) library from @tpolecat, and added a build step to check that the microsite compiles. Any code change that breaks the documentation will trigger a build failure.

 I was inspired by the [the new Cats microsite](https://typelevel.org/cats), and a long-slumbering urge to write lots of excellent documentation for algebird's traits and data structures. Tut and the new microsite are the perfect foundation for small, crisp, code-reviewable documentation contributions for existing and new data structures.

The PR also:
- Adds a "Resources for Learners" section to the microsite with links to every talk, paper, tutorial and set of slides I could find on Algebird.
- Bumps the `sbt-mima-plugin` to `0.1.11`, fixing this issue that was biting me on this PR and the #568 PR: https://github.com/typesafehub/migration-manager/issues/112
- Adds a bare-bones "cookbook" section. The plan is to add tutorials here from an "You want to get X done? Try these data structures!" viewpoint, vs the typical "Here's what this data structure can do for you!" approach we've taken.

### How to run the microsite
I couldn't figure out how to get this to push to the github pages section of my local fork, so you'll have to run the site yourself to give it a look.

Here's the magical incantation:

```
brew install jekyll
cd $algebird_home_dir
sbt docs/makeMicrosite
cd docs/target/site
jekyll serve
```

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,15 +1,25 @@
-Recommendations and requirements for how to best contribute to Algebird. 
+---
```
Comment (sritchie) [inline CONTRIBUTING.md]: this page gets copied over as a microsite page now.

File: .travis.yml [hunk 2]
```diff
@@ -6,10 +6,19 @@ matrix:
       script: ./sbt ++$TRAVIS_SCALA_VERSION clean test
 
     - scala: 2.11.8
-      script: ./sbt ++$TRAVIS_SCALA_VERSION clean coverage test coverageReport mimaReportBinaryIssues
+      script: ./sbt ++$TRAVIS_SCALA_VERSION clean coverage test coverageReport mimaReportBinaryIssues docs/makeMicrosite
```
Comment (sritchie) [inline .travis.yml]: this is required to check that all documentation still builds; `tut` actually compiles all the docs(!!), so regressions to the docs will cause a build failure here.

File: .travis.yml [hunk 3]
```diff
@@ -6,10 +6,19 @@ matrix:
...
       after_success:
         - bash <(curl -s https://codecov.io/bash)
 
+before_install:
+  - export PATH=${PATH}:./vendor/bundle
+
+install:
+  - rvm use 2.2.3 --install --fuzzy
```
Comment (sritchie) [inline .travis.yml]: required to build the docs microsite and generate jekyll.

File: README.md [hunk 4]
```diff
@@ -1,12 +1,16 @@
-## Algebird [![Build status](https://img.shields.io/travis/twitter/algebird/develop.svg)](http://travis-ci.org/twitter/algebird)
+## Algebird
+
+[![Build status](https://img.shields.io/travis/twitter/algebird/develop.svg)](http://travis-ci.org/twitter/algebird)
 [![Codecov branch](https://img.shields.io/codecov/c/github/twitter/algebird/develop.svg?maxAge=2592000)](https://codecov.io/github/twitter/algebird)
+[![Latest version](https://index.scala-lang.org/typelevel/cats/cats-core/latest.svg?color=orange)](https://index.scala-lang.org/twitter/algebird/algebird-core)
```
Comment (sritchie) [inline README.md]: Love this badge.

File: README.md [hunk 5]
```diff
@@ -24,33 +28,30 @@ import com.twitter.algebird.Operators._
...
-For example, average, moving average, max/min, set
-  union, approximate set size (in much less memory with HyperLogLog), approximate item counting
-  (using CountMinSketch).
-* All of these combine naturally in tuples, vectors, maps, options and more standard scala classes.
-* Implementations of Monoids for interesting approximation algorithms, such as Bloom filter,
-  HyperLogLog and CountMinSketch. These allow you to think of these sophisticated operations like
-  you might numbers, and add them up in hadoop or online to produce powerful statistics and
-  analytics.
+
+In the above, the class `Max[T]` signifies that the `+` operator should actually be `max` (this is accomplished by providing an implicit instance of a typeclass for `Max` that handles `+`).
```
Comment (sritchie) [inline README.md]: all I did here was undo the line wraps.

Comment (sritchie) [inline README.md]: andddd add backticks around code.

File: README.md [hunk 6]
```diff
@@ -24,33 +28,30 @@ import com.twitter.algebird.Operators._
...
+
+- Model a wide class of "reductions" as a sum on some iterator of a particular value type. For example, average, moving average, max/min, set union, approximate set size (in much less memory with HyperLogLog), approximate item counting (using CountMinSketch).
+- All of these combine naturally in tuples, vectors, maps, options and more standard scala classes.
+- Implementations of Monoids for interesting approximation algorithms, such as Bloom filter, HyperLogLog and CountMinSketch. These allow you to think of these sophisticated operations like you might numbers, and add them up in hadoop or online to produce powerful statistics and analytics.
 
 ## Documentation
 
-To learn more and find links to tutorials and information around the web, check out the [Algebird Wiki](https://github.com/twitter/algebird/wiki).
+To learn more and find links to tutorials and information around the web, check out the [website](https://twitter.github.io/algebird).
```
Comment (sritchie) [inline README.md]: link to the new website.

File: README.md [hunk 7]
```diff
@@ -24,33 +28,30 @@ import com.twitter.algebird.Operators._
...
 
-The latest ScalaDocs are hosted on Algebird's [Github Project Page](http://twitter.github.io/algebird).
+The latest API docs are hosted on Algebird's [ScalaDoc index](http://twitter.github.io/algebird/api/#index).
```
Comment (sritchie) [inline README.md]: same here.

File: README.md [hunk 8]
```diff
@@ -74,19 +75,14 @@ Current published artifacts are
 
 The suffix denotes the scala version.
 
-## Questions
-> Why not use spire?
-
-We didn't know about it when we started this code, but it seems like we're more focused on
-large scale analytics.
+## Projects using Algebird
```
Comment (sritchie) [inline README.md]: moved these here from the wiki.

File: build.sbt [hunk 9]
```diff
@@ -20,6 +22,24 @@ def scalaBinaryVersion(scalaVersion: String) = scalaVersion match {
...
+/**
+  * Remove 2.10 projects from doc generation, as the macros used in the projects
+  * cause problems generating the documentation on scala 2.10. As the APIs for 2.10
+  * and 2.11 are the same this has no effect on the resultant documentation, though
+  * it does mean that the scaladocs cannot be generated when the build is in 2.10 mode.
+  */
+def docsSourcesAndProjects(sv: String): (Boolean, Seq[ProjectReference]) =
+  CrossVersion.partialVersion(sv) match {
+    case Some((2, 10)) => (false, Nil)
+    case _ => (true, Seq(
```
Comment (sritchie) [inline build.sbt]: all projects in here get API docs for 2.11 and up.

File: build.sbt [hunk 10]
```diff
@@ -207,3 +231,54 @@ lazy val algebirdSpark = module("spark").settings(
...
+lazy val docs = project
+  .enablePlugins(MicrositesPlugin)
+  .settings(moduleName := "algebird-docs")
+  .settings(sharedSettings)
+  .settings(noPublishSettings)
+  .settings(unidocSettings)
+  .settings(ghpages.settings)
+  .settings(docSettings)
+  .settings(tutScalacOptions ~= (_.filterNot(Set("-Ywarn-unused-import", "-Ywarn-dead-code"))))
+  .dependsOn(algebirdCore)
```
Comment (sritchie) [inline build.sbt]: for now we only depend on core to generate tutorial stuff... users can add if they want to write docs for the non-core projects.

File: docs/src/main/tut/datatypes/repl.md [hunk 11]
```diff
@@ -0,0 +1,367 @@
+---
+layout: docs
+title:  "Learning Algebird at the REPL"
+section: "data"
+---
+
+# Learning Algebird Monoids at the REPL
```
Comment (sritchie) [inline docs/src/main/tut/datatypes/repl.md]: this is the "REPL" page from the wiki, with a nice new index, converted to `tut`.

File: docs/src/main/tut/index.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
+---
```
Comment (sritchie) [inline docs/src/main/tut/index.md]: this is a modified and trimmed version of the README, converted to `tut`.

File: docs/src/main/tut/faq.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
+---
+layout: page
```
Comment (sritchie) [inline docs/src/main/tut/faq.md]: copied from the wiki.

File: docs/src/main/tut/resources_for_learners.md [hunk 14]
```diff
@@ -0,0 +1,33 @@
+---
+layout: page
+title:  "Resources for Learners"
```
Comment (sritchie) [inline docs/src/main/tut/resources_for_learners.md]: I gathered as many of these as I could from the web.

File: docs/src/main/tut/typeclasses/abstract_algebra.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
+# Abstract Algebra
+
+Before getting into the details of Algebird, let's go through a few simple abstract algebra definitions.  Folks interested in more details may wish to refer to the excellent [*Notes on Abstract Algebra*](http://www.math.usm.edu/perry/old_classes/mat423fa11/notes_25aug2011.pdf) from University of Southern Mississipi professor John Perry.
```
Comment (sritchie) [inline docs/src/main/tut/typeclasses/abstract_algebra.md]: copied this from the wiki and converted `*` to `+` in the semigroup and monoid sections.

File: project/DocGen.scala [hunk 16]
```diff
@@ -1,43 +0,0 @@
-package algebird
```
Comment (sritchie) [inline project/DocGen.scala]: no longer needed!

File: project/plugins.sbt [hunk 17]
```diff
@@ -4,11 +4,14 @@ resolvers ++= Seq(
     url("http://dl.bintray.com/content/sbt/sbt-plugin-releases"))(Resolver.ivyStylePatterns)
 )
 
+addSbtPlugin("com.eed3si9n"       % "sbt-unidoc"      % "0.3.3")
+addSbtPlugin("com.fortysevendeg"  % "sbt-microsites"  % "0.3.0")
 addSbtPlugin("com.github.gseitz"  % "sbt-release"     % "1.0.0")
 addSbtPlugin("com.jsuereth"       % "sbt-pgp"         % "1.0.0")
-addSbtPlugin("com.typesafe"       % "sbt-mima-plugin" % "0.1.10")
+addSbtPlugin("com.typesafe"       % "sbt-mima-plugin" % "0.1.11")
```
Comment (sritchie) [inline project/plugins.sbt]: this fixes a bug with `IndexOutOfBoundsException` in the tests.

File: README.md (hunk 4 above)
Comment (ianoc-stripe) [inline README.md]: This looks wrong, it says cats-core on the page not algebird core

**Comment by ianoc** [comment]
This is very cool

File: README.md [hunk 18]
```diff
@@ -24,33 +28,30 @@ import com.twitter.algebird.Operators._
...
 Issues should be reported on the [GitHub issue tracker](https://github.com/twitter/algebird/issues).
 
-We use a lightweight form of project governence inspired by the one used by Apache projects.
+We use a lightweight form of project governance inspired by the one used by Apache projects.
+
 Please see [Contributing and Committership](https://github.com/twitter/analytics-infra-governance#contributing-and-committership) for our code of conduct and our pull request review process.
+
 The TL;DR is send us a pull request, iterate on the feedback + discussion, and get a +1 from a [Committer](COMMITTERS.md) in order to get your PR accepted.
 
 The current list of active committers (who can +1 a pull request) can be found here: [Committers](COMMITTERS.md)
```
Comment (ianoc-stripe) [inline README.md]: This link doesn't work to the Committers

File: docs/src/main/tut/datatypes/repl.md (hunk 11 above)
Comment (ianoc) [inline docs/src/main/tut/datatypes/repl.md]: Be good to link to this from the parent page, shows up in the index, but be good to be easier to find

**Comment by ianoc** [comment]
Seems like a solid thing, love the test/ci integration too. Some tiny nits. But seems like the thing thats worth just iterating on while on develop.

**Comment by ianoc** [comment]
LGTM

**Comment by tpolecat** [comment]
Hey, awesome that you're embracing checked documentation!

**Comment by sritchie** [comment]
@tpolecat keep the kool-aid coming... happy to drink more! So good

**Comment by johnynek** [comment]
Thanks a ton, @sritchie ! Can't wait to get these pushed to the github site.

**Comment by sritchie** [comment]
@johnynek Already done! twitter.github.io/algebird.
